aiohttp/payload.py

```diff
@@ -328,15 +400,165 @@ def __init__(
+
         try:
-            chunk = await loop.run_in_executor(None, self._value.read, 2**16)
+            # Get initial data and available length
+            available_len, chunk = await loop.run_in_executor(
+                None, self._read_and_available_len, remaining_content_len
+            )
+            # Process data chunks until done
             while chunk:
-                await writer.write(chunk)
-                chunk = await loop.run_in_executor(None, self._value.read, 2**16)
+                chunk_len = len(chunk)
+
+                # Write data with or without length constraint
+                if remaining_content_len is None:
+                    await writer.write(chunk)
+                else:
+                    await writer.write(chunk[:remaining_content_len])
+                    remaining_content_len -= chunk_len
+
+                total_written_len += chunk_len
+
+                # Check if we're done writing
+                if self._should_stop_writing(
+                    available_len, total_written_len, remaining_content_len
+                ):
+                    return
+
+                # Read next chunk
+                chunk = await loop.run_in_executor(
+                    None, self._read, remaining_content_len
+                )
         finally:
-            await loop.run_in_executor(None, self._value.close)
+            # Handle closing the file without awaiting to prevent cancellation issues
+            # when the StreamReader reaches EOF
+            self._schedule_file_close(loop)
+
+    def _should_stop_writing(
+        self,
+        available_len: Optional[int],
+        total_written_len: int,
+        remaining_content_len: Optional[int],
+    ) -> bool:
+        """
+        Determine if we should stop writing data.
+
+        Args:
+            available_len: Known size of the payload if available (None if unknown)
+            total_written_len: Number of bytes already written
+            remaining_content_len: Remaining bytes to be written for content-length limited responses
+
+        Returns:
+            True if we should stop writing data, based on either:
+            - Having written all available data (when size is known)
+            - Having written all requested content (when content-length is specified)
+
+        """
+        return (available_len is not None and total_written_len >= available_len) or (
+            remaining_content_len is not None and remaining_content_len <= 0
+        )
+
+    def _schedule_file_close(self, loop: asyncio.AbstractEventLoop) -> None:
+        """Schedule file closing without awaiting to prevent cancellation issues."""
+        close_future = loop.run_in_executor(None, self._value.close)
+        # Hold a strong reference to the future to prevent it from being
+        # garbage collected before it completes.
+        _CLOSE_FUTURES.add(close_future)
+        close_future.add_done_callback(_CLOSE_FUTURES.remove)
```

Comment:
There is a risk here that we lose exceptions as we never await on it.

Reply:
I think its an acceptable risk as `.close()` is very unlikely to fail. Its the same pattern we have in web_fileresponse which has worked well

Reply:
Since we are reading there isn't any buffer to flush, and the only reason it would fail is if file descriptors somehow get mixed up in Python, and at that point everything is likely broken